worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerPublisherImpl.java

```diff
@@ -66,18 +86,47 @@ public void handlePublish(byte[] data, String routingKey, RabbitTaskInformation
     {
         try {
             LOG.debug("Publishing message to {} with ack id {}", routingKey, taskInformation.getInboundMessageId());
+            final var publishHeaders = new HashMap<>(headers);
+            // Remove any previous dehydration id
+            publishHeaders.remove(RABBIT_HEADER_CAF_DEHYDRATION_ID);
+            final var outboundByteArray = getOutboundByteArray(data, routingKey, publishHeaders);
             AMQP.BasicProperties.Builder builder = new AMQP.BasicProperties().builder();
-            builder.headers(headers);
+            builder.headers(publishHeaders);
             builder.contentType("text/plain");
             builder.deliveryMode(2);
-
+            
             confirmListener.registerResponseSequence(channel.getNextPublishSeqNo(), taskInformation);
-            channel.basicPublish("", routingKey, builder.build(), data);
+            channel.basicPublish("", routingKey, builder.build(), outboundByteArray);
             metrics.incrementPublished();
-        } catch (IOException e) {
+        } catch (final IOException | QueueException e) {
             LOG.error("Failed to publish result of message {} to queue {}, rejecting", taskInformation.getInboundMessageId(), routingKey, e);
             metrics.incremementErrors();
             consumerEvents.add(new ConsumerRejectEvent(Long.valueOf(taskInformation.getInboundMessageId())));
         }
     }
+
+    private boolean shouldStoreTaskMessage(final int taskMessageSize) {
+        return config.getIsDehydrationEnabled() && taskMessageSize > config.getDehydrationThreshold();
+    }
+
+    private byte[] getOutboundByteArray(
+        final byte[] taskMessage,
+        final String routingKey,
+        final Map<String, Object> headers
+    ) throws QueueException {
+        try {
+            if (shouldStoreTaskMessage(taskMessage.length)) {
+                final TaskMessage outboundTaskMessage = codec.deserialise(taskMessage, TaskMessage.class);
```

Comment:
Change handlePublish to accept TaskMessage instead of byte[] so we avoid deserialization here.

Reply:
actually, we have to either deserialize or serialize here anyway, both storing and publishing require one or other of the actions.